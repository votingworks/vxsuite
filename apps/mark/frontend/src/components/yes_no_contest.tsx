import React, {
  useCallback,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react';
import {
  YesNoVote,
  OptionalYesNoVote,
  YesNoContest as YesNoContestInterface,
  YesOrNo,
  getContestDistrictName,
} from '@votingworks/types';
import {
  Button,
  ContestChoiceButton,
  DisplayTextForYesOrNo,
  Main,
  Modal,
  Prose,
  TextWithLineBreaks,
} from '@votingworks/ui';

import { getSingleYesNoVote } from '@votingworks/utils';
import { assert, Optional } from '@votingworks/basics';
import { ScrollDirections, UpdateVoteFunction } from '../config/types';

import { BallotContext } from '../contexts/ballot_context';

import { FONT_SIZES } from '../config/globals';

import {
  ContentHeader,
  ContestFooter,
  DistrictName,
  VariableContentContainer,
  ScrollControls,
  ScrollContainer,
  ScrollableContentWrapper,
  ChoicesGrid,
  ContestDescription,
} from './contest_screen_layout';

interface Props {
  contest: YesNoContestInterface;
  vote: OptionalYesNoVote;
  updateVote: UpdateVoteFunction;
}

export function YesNoContest({
  contest,
  vote,
  updateVote,
}: Props): JSX.Element {
  const { userSettings, electionDefinition } = useContext(BallotContext);
  assert(electionDefinition);
  const { election } = electionDefinition;
  const districtName = getContestDistrictName(election, contest);
  const scrollContainer = useRef<HTMLDivElement>(null);

  const [isScrollable, setIsScrollable] = useState(false);
  const [isScrollAtBottom, setIsScrollAtBottom] = useState(true);
  const [isScrollAtTop, setIsScrollAtTop] = useState(true);
  const [overvoteSelection, setOvervoteSelection] =
    useState<Optional<YesOrNo>>();
  const [deselectedVote, setDeselectedVote] = useState('');

  useEffect(() => {
    if (deselectedVote !== '') {
      const timer = setTimeout(() => setDeselectedVote(''), 100);
      return () => clearTimeout(timer);
    }
  }, [deselectedVote]);

  const updateContestChoicesScrollStates = useCallback(() => {
    const target = scrollContainer.current;
    /* istanbul ignore next - `target` should always exist, but sometimes it doesn't. Don't know how to create this condition in testing.  */
    if (!target) {
      return;
    }
    const targetMinHeight = FONT_SIZES[userSettings.textSize] * 8; // magic number: room for buttons + spacing
    const windowsScrollTopOffsetMagicNumber = 1; // Windows Chrome is often 1px when using scroll buttons.
    const windowsScrollTop = Math.ceil(target.scrollTop); // Windows Chrome scrolls to sub-pixel values.
    setIsScrollable(
      /* istanbul ignore next: Tested by Cypress */
      target.scrollHeight > target.offsetHeight &&
        /* istanbul ignore next: Tested by Cypress */
        target.offsetHeight > targetMinHeight
    );
    setIsScrollAtBottom(
      windowsScrollTop +
        target.offsetHeight +
        windowsScrollTopOffsetMagicNumber >= // Windows Chrome "gte" check.
        target.scrollHeight
    );
    setIsScrollAtTop(target.scrollTop === 0);
  }, [scrollContainer, userSettings.textSize]);

  const voteLength = vote?.length;
  useEffect(() => {
    updateContestChoicesScrollStates();
    window.addEventListener('resize', updateContestChoicesScrollStates);
    return () => {
      window.removeEventListener('resize', updateContestChoicesScrollStates);
    };
  }, [voteLength, updateContestChoicesScrollStates]);

  function handleUpdateSelection(newVote: YesOrNo) {
    if ((vote as string[] | undefined)?.includes(newVote)) {
      updateVote(contest.id, undefined);
      setDeselectedVote(newVote);
    } else {
      updateVote(contest.id, [newVote] as YesNoVote);
    }
  }

  function handleChangeVoteAlert(newValue: YesOrNo) {
    setOvervoteSelection(newValue);
  }

  /* istanbul ignore next: Tested by Cypress */
  function scrollContestChoices(direction: ScrollDirections) {
    const sc = scrollContainer.current;
    assert(sc);
    const currentScrollTop = sc.scrollTop;
    const { offsetHeight } = sc;
    const { scrollHeight } = sc;
    const idealScrollDistance = Math.round(offsetHeight * 0.75);
    const maxScrollableDownDistance =
      scrollHeight - offsetHeight - currentScrollTop;
    const maxScrollTop =
      direction === 'down'
        ? currentScrollTop + maxScrollableDownDistance
        : currentScrollTop;
    const idealScrollTop =
      direction === 'down'
        ? currentScrollTop + idealScrollDistance
        : currentScrollTop - idealScrollDistance;
    const top = idealScrollTop > maxScrollTop ? maxScrollTop : idealScrollTop;
    sc.scrollTo({ behavior: 'smooth', left: 0, top });
  }

  function closeOvervoteAlert() {
    setOvervoteSelection(undefined);
  }

  return (
    <React.Fragment>
      <Main flexColumn>
        <ContentHeader id="contest-header">
          <Prose id="audiofocus">
            <h1 aria-label={`${districtName} ${contest.title}.`}>
              <DistrictName>{districtName}</DistrictName>
              {contest.title}
            </h1>
            <p>
              Vote <strong>Yes</strong> or <strong>No</strong>.
              <span className="screen-reader-only">
                {contest.description}
                To navigate through the contest choices, use the down button. To
                move to the next contest, use the right button.
              </span>
            </p>
          </Prose>
        </ContentHeader>
        <VariableContentContainer
          showTopShadow={!isScrollAtTop}
          showBottomShadow={!isScrollAtBottom}
        >
          <ScrollContainer
            ref={scrollContainer}
            onScroll={updateContestChoicesScrollStates}
          >
            <ScrollableContentWrapper isScrollable={isScrollable}>
              <ContestDescription>
                <Prose>
                  <TextWithLineBreaks text={contest.description} />
                </Prose>
              </ContestDescription>
            </ScrollableContentWrapper>
          </ScrollContainer>
          {
            /* istanbul ignore next: Tested by Cypress */ isScrollable && (
              <ScrollControls aria-hidden>
                <Button
                  className="scroll-up"
                  large
                  variant="primary"
                  aria-hidden
                  value="up"
                  disabled={isScrollAtTop}
                  onPress={scrollContestChoices}
                >
                  <span>See More</span>
                </Button>
                <Button
                  className="scroll-down"
                  large
                  variant="primary"
                  aria-hidden
                  value="down"
                  disabled={isScrollAtBottom}
                  onPress={scrollContestChoices}
                >
                  <span>See More</span>
                </Button>
              </ScrollControls>
            )
          }
        </VariableContentContainer>
        <ContestFooter>
          <ChoicesGrid data-testid="contest-choices">
            {[
              { label: 'Yes', vote: 'yes' } as const,
              { label: 'No', vote: 'no' } as const,
            ].map((answer) => {
              const isChecked = getSingleYesNoVote(vote) === answer.vote;
              const isDisabled = !isChecked && !!vote;
              function handleDisabledClick() {
                handleChangeVoteAlert(answer.vote);
              }
              let prefixAudioText = '';
              if (isChecked) {
                prefixAudioText = 'Selected,';
              } else if (deselectedVote === answer.vote) {
                prefixAudioText = 'Deselected,';
              }
              return (
                <ContestChoiceButton
                  key={answer.vote}
                  choice={answer.vote}
                  isSelected={isChecked}
                  onPress={
                    isDisabled ? handleDisabledClick : handleUpdateSelection
                  }
                  ariaLabel={`${prefixAudioText} ${answer.label} on ${contest.title}`}
                  label={answer.label}
                />
              );
            })}
          </ChoicesGrid>
        </ContestFooter>
      </Main>
      {overvoteSelection && (
        <Modal
          centerContent
          content={
            <Prose>
              {overvoteSelection && (
                <p id="modalaudiofocus">
                  Do you want to change your vote to{' '}
                  <strong>{DisplayTextForYesOrNo[overvoteSelection]}</strong>?
                  To change your vote, first unselect your vote for{' '}
                  <strong>
                    {
                      {
                        no: DisplayTextForYesOrNo.yes,
                        yes: DisplayTextForYesOrNo.no,
                      }[overvoteSelection]
                    }
                  </strong>
                  .
                </p>
              )}
            </Prose>
          }
          actions={
            <Button variant="primary" autoFocus onPress={closeOvervoteAlert}>
              Okay
            </Button>
          }
        />
      )}
    </React.Fragment>
  );
}
