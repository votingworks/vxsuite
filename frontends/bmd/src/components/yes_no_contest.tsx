import React, {
  useCallback,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react';
import {
  YesNoVote,
  OptionalYesNoVote,
  YesNoContest as YesNoContestInterface,
  Optional,
  YesOrNo,
} from '@votingworks/types';
import { Button, Main } from '@votingworks/ui';

import { assert, getSingleYesNoVote } from '@votingworks/utils';
import {
  EventTargetFunction,
  ScrollDirections,
  UpdateVoteFunction,
} from '../config/types';

import { BallotContext } from '../contexts/ballot_context';

import { FONT_SIZES, YES_NO_VOTES } from '../config/globals';
import { ChoiceButton } from './choice_button';
import { Modal } from './modal';
import { Prose } from './prose';
import { Text, TextWithLineBreaks } from './text';
import {
  ContentHeader,
  ContestFooter,
  ContestSection,
  VariableContentContainer,
  ScrollControls,
  ScrollContainer,
  ScrollableContentWrapper,
  ChoicesGrid,
} from './contest_screen_layout';

interface Props {
  contest: YesNoContestInterface;
  vote: OptionalYesNoVote;
  updateVote: UpdateVoteFunction;
}

export function YesNoContest({
  contest,
  vote,
  updateVote,
}: Props): JSX.Element {
  const { userSettings } = useContext(BallotContext);
  const scrollContainer = useRef<HTMLDivElement>(null);

  const [isScrollable, setIsScrollable] = useState(false);
  const [isScrollAtBottom, setIsScrollAtBottom] = useState(true);
  const [isScrollAtTop, setIsScrollAtTop] = useState(true);
  const [overvoteSelection, setOvervoteSelection] = useState<
    Optional<YesOrNo>
  >();
  const [deselectedVote, setDeselectedVote] = useState('');

  useEffect(() => {
    if (deselectedVote !== '') {
      const timer = setTimeout(() => setDeselectedVote(''), 100);
      return () => clearTimeout(timer);
    }
  }, [deselectedVote]);

  const updateContestChoicesScrollStates = useCallback(() => {
    const target = scrollContainer.current;
    /* istanbul ignore next - `target` should always exist, but sometimes it doesn't. Don't know how to create this condition in testing.  */
    if (!target) {
      return;
    }
    const targetMinHeight = FONT_SIZES[userSettings.textSize] * 8; // magic number: room for buttons + spacing
    const windowsScrollTopOffsetMagicNumber = 1; // Windows Chrome is often 1px when using scroll buttons.
    const windowsScrollTop = Math.ceil(target.scrollTop); // Windows Chrome scrolls to sub-pixel values.
    setIsScrollable(
      /* istanbul ignore next: Tested by Cypress */
      target.scrollHeight > target.offsetHeight &&
        /* istanbul ignore next: Tested by Cypress */
        target.offsetHeight > targetMinHeight
    );
    setIsScrollAtBottom(
      windowsScrollTop +
        target.offsetHeight +
        windowsScrollTopOffsetMagicNumber >= // Windows Chrome "gte" check.
        target.scrollHeight
    );
    setIsScrollAtTop(target.scrollTop === 0);
  }, [scrollContainer, userSettings.textSize]);

  const voteLength = vote?.length;
  useEffect(() => {
    updateContestChoicesScrollStates();
    window.addEventListener('resize', updateContestChoicesScrollStates);
    return () => {
      window.removeEventListener('resize', updateContestChoicesScrollStates);
    };
  }, [voteLength, updateContestChoicesScrollStates]);

  const handleUpdateSelection: EventTargetFunction = (event) => {
    const newVote = (event.currentTarget as HTMLInputElement).dataset
      .choice as YesOrNo;
    if ((vote as string[] | undefined)?.includes(newVote)) {
      updateVote(contest.id, undefined);
      setDeselectedVote(newVote);
    } else {
      updateVote(contest.id, [newVote] as YesNoVote);
    }
  };

  function handleChangeVoteAlert(newValue: YesOrNo) {
    setOvervoteSelection(newValue);
  }

  /* istanbul ignore next: Tested by Cypress */
  const scrollContestChoices: EventTargetFunction = (event) => {
    const direction = (event.target as HTMLElement).dataset
      .direction as ScrollDirections;
    const sc = scrollContainer.current;
    assert(sc);
    const currentScrollTop = sc.scrollTop;
    const { offsetHeight } = sc;
    const { scrollHeight } = sc;
    const idealScrollDistance = Math.round(offsetHeight * 0.75);
    const maxScrollableDownDistance =
      scrollHeight - offsetHeight - currentScrollTop;
    const maxScrollTop =
      direction === 'down'
        ? currentScrollTop + maxScrollableDownDistance
        : currentScrollTop;
    const idealScrollTop =
      direction === 'down'
        ? currentScrollTop + idealScrollDistance
        : currentScrollTop - idealScrollDistance;
    const top = idealScrollTop > maxScrollTop ? maxScrollTop : idealScrollTop;
    sc.scrollTo({ behavior: 'smooth', left: 0, top });
  };

  function closeOvervoteAlert() {
    setOvervoteSelection(undefined);
  }

  return (
    <React.Fragment>
      <Main>
        <ContentHeader id="contest-header">
          <Prose id="audiofocus">
            <h1 aria-label={`${contest.section} ${contest.title}.`}>
              <ContestSection>{contest.section}</ContestSection>
              {contest.title}
            </h1>
            <p>
              <strong>Vote Yes or No.</strong>
              <span className="screen-reader-only">
                {contest.description}
                To navigate through the contest choices, use the down button. To
                move to the next contest, use the right button.
              </span>
            </p>
          </Prose>
        </ContentHeader>
        <VariableContentContainer
          showTopShadow={!isScrollAtTop}
          showBottomShadow={!isScrollAtBottom}
        >
          <ScrollContainer
            ref={scrollContainer}
            onScroll={updateContestChoicesScrollStates}
          >
            <ScrollableContentWrapper isScrollable={isScrollable}>
              <Prose>
                <TextWithLineBreaks text={contest.description} />
              </Prose>
            </ScrollableContentWrapper>
          </ScrollContainer>
          {
            /* istanbul ignore next: Tested by Cypress */ isScrollable && (
              <ScrollControls aria-hidden>
                <Button
                  className="scroll-up"
                  large
                  primary
                  aria-hidden
                  data-direction="up"
                  disabled={isScrollAtTop}
                  onPress={scrollContestChoices}
                >
                  <span>See More</span>
                </Button>
                <Button
                  className="scroll-down"
                  large
                  primary
                  aria-hidden
                  data-direction="down"
                  disabled={isScrollAtBottom}
                  onPress={scrollContestChoices}
                >
                  <span>See More</span>
                </Button>
              </ScrollControls>
            )
          }
        </VariableContentContainer>
        <ContestFooter>
          <ChoicesGrid data-testid="contest-choices">
            {[
              { label: 'Yes', vote: 'yes' } as const,
              { label: 'No', vote: 'no' } as const,
            ].map((answer) => {
              const isChecked = getSingleYesNoVote(vote) === answer.vote;
              const isDisabled = !isChecked && !!vote;
              function handleDisabledClick() {
                handleChangeVoteAlert(answer.vote);
              }
              let prefixAudioText = '';
              if (isChecked) {
                prefixAudioText = 'Selected,';
              } else if (deselectedVote === answer.vote) {
                prefixAudioText = 'Deselected,';
              }
              return (
                <ChoiceButton
                  key={answer.vote}
                  choice={answer.vote}
                  isSelected={isChecked}
                  onPress={
                    isDisabled ? handleDisabledClick : handleUpdateSelection
                  }
                >
                  <Prose>
                    <Text
                      aria-label={`${prefixAudioText} ${answer.label} on ${
                        contest.shortTitle ?? contest.title
                      }`}
                      wordBreak
                    >
                      {answer.label}
                    </Text>
                  </Prose>
                </ChoiceButton>
              );
            })}
          </ChoicesGrid>
        </ContestFooter>
      </Main>
      {overvoteSelection && (
        <Modal
          centerContent
          content={
            <Prose>
              {overvoteSelection && (
                <p id="modalaudiofocus">
                  Do you want to change your vote to{' '}
                  <strong>{YES_NO_VOTES[overvoteSelection]}</strong>? To change
                  your vote, first unselect your vote for{' '}
                  <strong>
                    {
                      {
                        no: YES_NO_VOTES.yes,
                        yes: YES_NO_VOTES.no,
                      }[overvoteSelection]
                    }
                  </strong>
                  .
                </p>
              )}
            </Prose>
          }
          actions={
            <Button primary autoFocus onPress={closeOvervoteAlert}>
              Okay
            </Button>
          }
        />
      )}
    </React.Fragment>
  );
}
