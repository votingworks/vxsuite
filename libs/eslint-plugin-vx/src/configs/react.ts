import recommended = require('./recommended');

const jsExtensions = ['.js', '.jsx'];
const tsExtensions = ['.ts', '.tsx'];
const allExtensions = jsExtensions.concat(tsExtensions);

export = {
  ...recommended,
  extends: [
    ...recommended.extends
      .map((name) => (name === 'airbnb-base' ? 'airbnb' : name))
      .filter((name) => name !== 'plugin:prettier/recommended'),
    'airbnb/hooks',
    'plugin:react/recommended',
    'plugin:jsx-a11y/recommended',
    'plugin:prettier/recommended', // Enables eslint-plugin-prettier and displays prettier errors as ESLint errors. Make sure this is always the last configuration in the extends array.
  ],
  settings: {
    ...recommended.settings,
    react: {
      version: 'detect',
    },
  },
  rules: {
    ...recommended.rules,
    'jsx-a11y/label-has-associated-control': [
      'error',
      {
        controlComponents: ['Select'],
      },
    ],
    'react/jsx-filename-extension': ['error', { extensions: allExtensions }],
    'react/jsx-fragments': ['error', 'element'],
    'react/jsx-no-bind': 'off',
    'react/jsx-no-constructed-context-values': 'off',
    'react/jsx-no-useless-fragment': ['error', { allowExpressions: true }],
    'react/jsx-one-expression-per-line': 'off',
    'react/jsx-props-no-spreading': 'off',
    'react/jsx-wrap-multilines': 'off',
    'react/prop-types': 'off',
    'react/require-default-props': 'off',
  },
};
