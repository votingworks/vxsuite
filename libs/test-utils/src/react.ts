/* eslint-disable no-console -- this file monkey-patches `console` */

/**
 * Suppresses the "Can't perform a React state update on an unmounted component" warning by
 * modifying the `console.error` method to ignore the warning.
 *
 * This is a temporary workaround until we upgrade to React 18.
 */
export function suppressReact17UnmountedWarning(): void {
  const originalConsoleError = console.error;
  console.error = (...args: unknown[]) => {
    if (
      typeof args[0] === 'string' &&
      args[0].startsWith(
        `Warning: Can't perform a React state update on an unmounted component`
      )
    ) {
      return;
    }

    originalConsoleError(...args);
  };
}
