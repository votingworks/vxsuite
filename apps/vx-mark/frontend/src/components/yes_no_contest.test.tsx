import React from 'react';
import { fireEvent, render, screen, within } from '@testing-library/react';
import {
  DistrictIdSchema,
  unsafeParse,
  YesNoContest as YesNoContestInterface,
} from '@votingworks/types';

import { YesNoContest } from './yes_no_contest';

const contest: YesNoContestInterface = {
  description:
    'Institute a garbage collection program that collects garbage on weekdays across the county.',
  districtId: unsafeParse(DistrictIdSchema, 'district-id'),
  id: 'contest-id',
  section: 'County',
  shortTitle: 'Prop 1',
  title: 'Prop 1: Garbage Collection Program',
  type: 'yesno',
};

describe('supports yes/no contest', () => {
  it('allows voting for both yes and no', () => {
    const updateVote = jest.fn();
    const { container } = render(
      <YesNoContest
        contest={contest}
        vote={undefined}
        updateVote={updateVote}
      />
    );
    expect(container).toMatchSnapshot();

    const contestChoices = screen.getByTestId('contest-choices');
    fireEvent.click(within(contestChoices).getByText('Yes').closest('button')!);
    expect(updateVote).toHaveBeenCalledTimes(1);

    fireEvent.click(within(contestChoices).getByText('No').closest('button')!);
    expect(updateVote).toHaveBeenCalledTimes(2);
  });

  it('displays warning when attempting to change vote', () => {
    const updateVote = jest.fn();
    const { container } = render(
      <YesNoContest contest={contest} vote={['yes']} updateVote={updateVote} />
    );
    expect(container).toMatchSnapshot();
    const contestChoices = screen.getByTestId('contest-choices');
    fireEvent.click(within(contestChoices).getByText('No').closest('button')!);
    expect(
      screen.getAllByText(
        (_, element) =>
          element?.textContent ===
          'Do you want to change your vote to No? To change your vote, first unselect your vote for Yes.'
      )
    ).toBeTruthy();
    fireEvent.click(screen.getByText('Okay'));
  });
});
