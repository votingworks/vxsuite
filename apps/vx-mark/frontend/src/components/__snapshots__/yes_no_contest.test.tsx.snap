// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`supports yes/no contest allows voting for both yes and no 1`] = `
.c8 {
  position: relative;
  border-radius: 0.125rem;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  background: #FFFFFF;
  cursor: pointer;
  padding: 0.5rem 0.5rem 0.5rem 4rem;
  text-align: left;
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

.c8::before {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  border-right: 1px solid;
  border-color: rgb(211,211,211);
  border-radius: 0.125rem 0 0 0.125rem;
  background: #ffffff;
  width: 3rem;
  text-align: center;
  color: #028099;
  font-size: 2rem;
  font-weight: 700;
  content: '';
}

.c0 {
  margin: 0 auto;
  width: 100%;
  padding: 1rem 30px 0.5rem;
}

.c6 {
  margin: 0 auto;
  width: 100%;
  padding: 30px 20px 0;
}

.c1 {
  text-transform: uppercase;
  font-size: 0.85rem;
  font-weight: 600;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  overflow: auto;
}

.c2::before,
.c2::after {
  position: absolute;
  z-index: 1;
  width: 100%;
  height: 0.25rem;
  content: '';
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c2::before {
  top: 0;
  opacity: 0;
  background: linear-gradient( to bottom, rgb(177,186,190) 0%, transparent 100% );
}

.c2::after {
  bottom: 0;
  opacity: 0;
  background: linear-gradient( to bottom, transparent 0%, rgb(177,186,190) 100% );
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.c4 {
  margin: 0 auto;
  width: 100%;
  padding: 0.25rem 20px 0;
}

.c5 {
  padding: 0 10px;
}

.c7 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 20px;
  margin-bottom: 40px;
}

@media (min-width:480px) {
  .c8 {
    padding: 1rem 1rem 1rem 4rem;
  }
}

<div>
  <main
    class="sc-kLwgWK kRCmMs"
  >
    <div
      class="c0"
      id="contest-header"
    >
      <div
        class="sc-hKwCoD gmiWky"
        id="audiofocus"
      >
        <h1
          aria-label="County Prop 1: Garbage Collection Program."
        >
          <div
            class="c1"
          >
            County
          </div>
          Prop 1: Garbage Collection Program
        </h1>
        <p>
          Vote 
          <strong>
            Yes
          </strong>
           or 
          <strong>
            No
          </strong>
          .
          <span
            class="screen-reader-only"
          >
            Institute a garbage collection program that collects garbage on weekdays across the county.
            To navigate through the contest choices, use the down button. To move to the next contest, use the right button.
          </span>
        </p>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <div
              class="sc-hKwCoD gmiWky"
            >
              <p
                class="sc-bdvvaa NLRzK"
              >
                <span>
                  Institute a garbage collection program that collects garbage on weekdays across the county.
                </span>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c6"
    >
      <div
        class="c7"
        data-testid="contest-choices"
      >
        <button
          class="c8"
          data-choice="yes"
          data-selected="false"
          role="option"
          type="button"
        >
          <div
            class="sc-hKwCoD gmiWky"
          >
            <p
              aria-label=" Yes on Prop 1"
              class="sc-bdvvaa eSpfPW"
            >
              Yes
            </p>
          </div>
        </button>
        <button
          class="c8"
          data-choice="no"
          data-selected="false"
          role="option"
          type="button"
        >
          <div
            class="sc-hKwCoD gmiWky"
          >
            <p
              aria-label=" No on Prop 1"
              class="sc-bdvvaa eSpfPW"
            >
              No
            </p>
          </div>
        </button>
      </div>
    </div>
  </main>
</div>
`;

exports[`supports yes/no contest displays warning when attempting to change vote 1`] = `
.c9 {
  position: relative;
  border-radius: 0.125rem;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  background: #FFFFFF;
  cursor: pointer;
  padding: 0.5rem 0.5rem 0.5rem 4rem;
  text-align: left;
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

.c9::before {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  border-right: 1px solid;
  border-color: rgb(211,211,211);
  border-radius: 0.125rem 0 0 0.125rem;
  background: #ffffff;
  width: 3rem;
  text-align: center;
  color: #028099;
  font-size: 2rem;
  font-weight: 700;
  content: '';
}

.c8 {
  position: relative;
  border-radius: 0.125rem;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  background: #028099;
  cursor: pointer;
  padding: 0.5rem 0.5rem 0.5rem 4rem;
  text-align: left;
  color: #FFFFFF;
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

.c8::before {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  border-right: 1px solid;
  border-color: #028099;
  border-radius: 0.125rem 0 0 0.125rem;
  background: #ffffff;
  width: 3rem;
  text-align: center;
  color: #028099;
  font-size: 2rem;
  font-weight: 700;
  content: 'âœ“';
}

.c0 {
  margin: 0 auto;
  width: 100%;
  padding: 1rem 30px 0.5rem;
}

.c6 {
  margin: 0 auto;
  width: 100%;
  padding: 30px 20px 0;
}

.c1 {
  text-transform: uppercase;
  font-size: 0.85rem;
  font-weight: 600;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  overflow: auto;
}

.c2::before,
.c2::after {
  position: absolute;
  z-index: 1;
  width: 100%;
  height: 0.25rem;
  content: '';
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c2::before {
  top: 0;
  opacity: 0;
  background: linear-gradient( to bottom, rgb(177,186,190) 0%, transparent 100% );
}

.c2::after {
  bottom: 0;
  opacity: 0;
  background: linear-gradient( to bottom, transparent 0%, rgb(177,186,190) 100% );
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.c4 {
  margin: 0 auto;
  width: 100%;
  padding: 0.25rem 20px 0;
}

.c5 {
  padding: 0 10px;
}

.c7 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 20px;
  margin-bottom: 40px;
}

@media (min-width:480px) {
  .c9 {
    padding: 1rem 1rem 1rem 4rem;
  }
}

@media (min-width:480px) {
  .c8 {
    padding: 1rem 1rem 1rem 4rem;
  }
}

<div>
  <main
    class="sc-kLwgWK kRCmMs"
  >
    <div
      class="c0"
      id="contest-header"
    >
      <div
        class="sc-hKwCoD gmiWky"
        id="audiofocus"
      >
        <h1
          aria-label="County Prop 1: Garbage Collection Program."
        >
          <div
            class="c1"
          >
            County
          </div>
          Prop 1: Garbage Collection Program
        </h1>
        <p>
          Vote 
          <strong>
            Yes
          </strong>
           or 
          <strong>
            No
          </strong>
          .
          <span
            class="screen-reader-only"
          >
            Institute a garbage collection program that collects garbage on weekdays across the county.
            To navigate through the contest choices, use the down button. To move to the next contest, use the right button.
          </span>
        </p>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <div
              class="sc-hKwCoD gmiWky"
            >
              <p
                class="sc-bdvvaa NLRzK"
              >
                <span>
                  Institute a garbage collection program that collects garbage on weekdays across the county.
                </span>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c6"
    >
      <div
        class="c7"
        data-testid="contest-choices"
      >
        <button
          class="c8"
          data-choice="yes"
          data-selected="true"
          role="option"
          type="button"
        >
          <div
            class="sc-hKwCoD gmiWky"
          >
            <p
              aria-label="Selected, Yes on Prop 1"
              class="sc-bdvvaa eSpfPW"
            >
              Yes
            </p>
          </div>
        </button>
        <button
          class="c9"
          data-choice="no"
          data-selected="false"
          role="option"
          type="button"
        >
          <div
            class="sc-hKwCoD gmiWky"
          >
            <p
              aria-label=" No on Prop 1"
              class="sc-bdvvaa eSpfPW"
            >
              No
            </p>
          </div>
        </button>
      </div>
    </div>
  </main>
</div>
`;
